pkg/reconciler/ingress/status.go

```diff
@@ -0,0 +1,301 @@
+	probeHost      string
+	virtualService *v1alpha3.VirtualService
+
+	// pendingCount is the number of pods that haven't been successfully probed yet
+	pendingCount int32
+	lastAccessed time.Time
+
+	context context.Context
+	cancel  func()
+}
+
+type workItem struct {
+	*probingState
+	podIP string
+}
+
+// StatusManager provides a way to check if a VirtualService is ready
+type StatusManager interface {
+	IsReady(vs *v1alpha3.VirtualService) (bool, error)
+}
+
+// StatusProber provides a way to check if a VirtualService is ready by probing the Envoy pods
+// handling that VirtualService.
+type StatusProber struct {
+	logger *zap.SugaredLogger
+
+	// mu guards probingStates
+	mu            sync.Mutex
+	probingStates map[string]*probingState
+
+	workQueue workqueue.RateLimitingInterface
+
+	gatewayLister    istiolisters.GatewayLister
+	podLister        corev1listers.PodLister
+	transportFactory func() http.RoundTripper
+
+	readyCallback func(*v1alpha3.VirtualService)
+
+	probeConcurrency int
+	stateExpiration  time.Duration
+	cleanupPeriod    time.Duration
+}
+
+// NewStatusProber creates a new instance of StatusProber
+func NewStatusProber(
+	logger *zap.SugaredLogger,
+	gatewayLister istiolisters.GatewayLister,
+	podLister corev1listers.PodLister,
+	transportFactory prober.TransportFactory,
+	readyCallback func(*v1alpha3.VirtualService)) *StatusProber {
+	return &StatusProber{
+		logger:        logger,
+		probingStates: make(map[string]*probingState),
+		workQueue: workqueue.NewNamedRateLimitingQueue(
+			workqueue.DefaultControllerRateLimiter(),
+			"ProbingQueue"),
+		gatewayLister:    gatewayLister,
+		podLister:        podLister,
+		transportFactory: transportFactory,
+		readyCallback:    readyCallback,
+		probeConcurrency: probeConcurrency,
+		stateExpiration:  stateExpiration,
+		cleanupPeriod:    cleanupPeriod,
+	}
+}
+
+// IsReady checks if the provided VirtualService is ready, i.e. the Envoy pods serving the VirtualService
+// have all been updated. This function is designed to be used by the ClusterIngress controller, i.e. it
+// will be called in the order of reconciliation. This means that if IsReady is called on a VirtualService,
+// this VirtualService is the latest know version and therefore anything related to older versions can be ignored.
+// Also, it means that IsReady is not called concurrently.
+func (m *StatusProber) IsReady(vs *v1alpha3.VirtualService) (bool, error) {
+	key := fmt.Sprintf("%s/%s", vs.Namespace, vs.Name)
+
+	// Find the probe host
+	var probeHost string
+	for _, host := range vs.Spec.Hosts {
+		if strings.HasSuffix(host, resources.ProbeHostSuffix) {
+			if probeHost != "" {
+				return false, fmt.Errorf("only one probe host can be defined in a VirtualService, found at least 2: %s, %s", probeHost, host)
+			}
+			probeHost = host
+		}
+	}
+	if probeHost == "" {
+		m.logger.Errorf("the provided VirtualService doesn't contain a probe host. Hosts: %v", vs.Spec.Hosts)
```

Comment:
```suggestion
		m.logger.Errorf("The provided VirtualService doesn't contain a probe host. Hosts: %v", vs.Spec.Hosts)
```
Probably want to include the target suffix?